src/webgl/loading.js

```diff
@@ -202,52 +219,51 @@ p5.prototype.loadModel = function(path) {
   return model;
 };
 
-function parseMtl(mtlPath){ //accepts mtlPath to load file
-  return new Promise((resolve,reject)=>{
-    // console.log('parser is called');
-    let currentMaterial = null;
-    p5.prototype.loadStrings(
-      mtlPath,
-      lines => {
-        for (let line = 0; line < lines.length; ++line){
-          const tokens = lines[line].trim().split(/\s+/);
-          if(tokens[0] === 'newmtl') {
-            const materialName = tokens[1];
-            currentMaterial = materialName;
-            materials[currentMaterial] = {};
-          }else if (tokens[0] === 'Kd'){
-            //Diffuse color
-            materials[currentMaterial].diffuseColor = [
-              parseFloat(tokens[1]),
-              parseFloat(tokens[2]),
-              parseFloat(tokens[3])
-            ];
-          } else if (tokens[0] === 'Ka'){
-            //Ambient Color
-            materials[currentMaterial].ambientColor = [
-              parseFloat(tokens[1]),
-              parseFloat(tokens[2]),
-              parseFloat(tokens[3])
-            ];
-          }else if (tokens[0] === 'Ks'){
-            //Specular color
-            materials[currentMaterial].specularColor = [
-              parseFloat(tokens[1]),
-              parseFloat(tokens[2]),
-              parseFloat(tokens[3])
-            ];
-
-          }else if (tokens[0] === 'map_Kd') {
-            //Texture path
-            materials[currentMaterial].texturePath = tokens[1];
-          }
+async function parseMtl(self,mtlPath){ //accepts mtlPath to load file
+  let materials= {};
+  let currentMaterial = null;
+  self.loadStrings(
```

Comment:
It looks like this function completes without waiting for the result of `loadStrings` now that it no longer returns a promise. So I think that means it will immediately return `{}`, and then asynchronously, properties will get added to it. I'm guessing it's important to have the complete materials object before continuing to parse .obj files later, so maybe it would make sense to add back in something like you had before?
```js
const materials = {};
const doneLoading = new Promise((resolve, reject) => {
  self.loadStrings(
    mtlPath,
    lines => {
      // do parsing and add to materials here
      resolve();
    },
    reject
  )
});
await doneLoading;
return materials;
```

Reply:
@davepagurek I changed my multiple mtl file support  functionality to this:
```
else if (fileType.match(/\.obj$/i)) {
    this.loadStrings(
      path,
      async lines => {
        const parsedMaterialPromises=[];
        try{
          for (let i = 0; i < lines.length; i++) {
            const mtllibMatch = lines[i].match(/^mtllib (.+)/);
            if (mtllibMatch) {
              let mtlPath='';
              const mtlFilename = mtllibMatch[1];
              const objPathParts = path.split('/');
              if(objPathParts.length > 1){
                objPathParts.pop();
                const objFolderPath = objPathParts.join('/');
                mtlPath = objFolderPath + '/' + mtlFilename;
              }else{
                mtlPath = mtlFilename;
              }
              const parsedMaterialsIndividual = await parseMtl(self, mtlPath);
              parsedMaterialPromises.push(parsedMaterialsIndividual);
            }
          }
          if(parsedMaterialPromises.length>0){
            const parsedMaterialsArray=
            await Promise.all(parsedMaterialPromises);
            const parsedMaterials = Object.assign({}, ...parsedMaterialsArray);
            await parseObj(model, lines, parsedMaterials);
          } }catch (error) {
          if (failureCallback) {
            failureCallback(error);
          } else {
            p5._friendlyError('Error during parsing: ' + error.message);
          }
        }
        finally{
          if (normalize) {
            model.normalize();
          }

          self._decrementPreload();
          if (typeof successCallback === 'function') {
            successCallback(model);
          }
        }
      },
      failureCallback
    );
  } else {
    p5._friendlyFileLoadError(3, path);
    if (failureCallback) {
      failureCallback();
    } else {
      p5._friendlyError(
        'Sorry, the file type is invalid. Only OBJ and STL files are supported.'
      );
    }
  }
  return model;
};
```

now single mtl files are loaded and then only passed to parseObj but multiple files still pass before fully loading all the materials thus the error: p5.js says: Error during parsing: Cannot read properties of undefined (reading 'diffuseColor')

In the earlier approach even after adding resolve, reject back and some other changes too, still the parseObj was called beforehand. 

where is the logic breaking in the current code that single mtl files are passed after materials fully loading but not multiple mtl files? I'm trying to still find since I used `Promise.all` also.

Reply:
A few little things:
- Right now you do `const parsedMaterialsIndividual = await parseMtl(self, mtlPath);`. By `await`ing this, this means synchronously waiting for the material to be loaded/parsed, and the returned value is the actual result and not a promise. This works, but it means all mtl files are loaded one after another, and the `Promise.all` has no effect due to the results not being promises any more. If you remove the `await` on this line, the result is a promise, and the code only waits for them to finish at the `Promise.all`, meaning they all get loaded in parallel, which seems more like what you want.
- It sounds like right now, `await`ing `parseMtl` doesn't fully wait for it to be loaded. What does your `parseMtl` implementation look like right now? In the first comment in this chain, I had added some suggestions on how to update `parseMtl` to await for loading to complete, have you tried any of that out yet?

Let me know if I can help elaborate on any of those points!

Reply:
@davepagurek below is the parseMtl function:
you had suggested using doneLoading and this is how I'm currently implementing the function:
```
function parseMtl(p5,mtlPath){ //accepts mtlPath to load file
  return new Promise((resolve, reject)=>{
    let currentMaterial = null;
    let materials= {};
    p5.loadStrings(
      mtlPath,
      lines => {
        for (let line = 0; line < lines.length; ++line){
          const tokens = lines[line].trim().split(/\s+/);
          if(tokens[0] === 'newmtl') {
            const materialName = tokens[1];
            currentMaterial = materialName;
            materials[currentMaterial] = {};
          }else if (tokens[0] === 'Kd'){
          //Diffuse color
            materials[currentMaterial].diffuseColor = [
              parseFloat(tokens[1]),
              parseFloat(tokens[2]),
              parseFloat(tokens[3])
            ];
          } else if (tokens[0] === 'Ka'){
          //Ambient Color
            materials[currentMaterial].ambientColor = [
              parseFloat(tokens[1]),
              parseFloat(tokens[2]),
              parseFloat(tokens[3])
            ];
          }else if (tokens[0] === 'Ks'){
          //Specular color
            materials[currentMaterial].specularColor = [
              parseFloat(tokens[1]),
              parseFloat(tokens[2]),
              parseFloat(tokens[3])
            ];

          }else if (tokens[0] === 'map_Kd') {
          //Texture path
            materials[currentMaterial].texturePath = tokens[1];
          }
        }
        resolve(materials);
      },reject
    );
  });
}
```